ratelimitstore.go

```diff
@@ -0,0 +1,77 @@
+package desync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
+)
+
+type ThrottleOptions struct {
+	eventRate                float64
+	burstRate int
+	timeout time.Duration
+	immediateOrFail bool
+}
+
+type RateLimitedLocalStore struct {
+
+	wrappedStore WriteStore
+
+	limiter  *rate.Limiter
+
+	options ThrottleOptions
+
+}
+
+var RateLimitedExceeded = errors.New("Rate Limit Exceeded")
+
+func NewRateLimitedLocalStore(s WriteStore, options ThrottleOptions) *RateLimitedLocalStore {
+	
+	limiter := rate.NewLimiter(rate.Limit(options.eventRate), options.burstRate)
+	return &RateLimitedLocalStore{wrappedStore: s,limiter: limiter, options: options }
+}
+
+func (s RateLimitedLocalStore) GetChunk(id ChunkID) (*Chunk, error) {
+
+	return s.wrappedStore.GetChunk(id)
+}
+
+func (s RateLimitedLocalStore) HasChunk(id ChunkID) (bool, error) {
+
+
+	return s.wrappedStore.HasChunk(id)
+}
+
+
+func (s RateLimitedLocalStore) StoreChunk(chunk *Chunk) error {
+	
+	// This isn't ideal because what I'm really interested is in size over the wire.
+	_, err := chunk.Data()
+	if err != nil {
+		return err
+	}
+	
+	//size := len(b)
+	ctx, cancel:= context.WithTimeout(context.Background(), s.options.timeout)
+	defer cancel()
+	
+	if s.options.immediateOrFail{
```

Comment:
Do you actually need `immediateOrFail` ? Seems unexpected to cause a failure because the rate-limit was exceeded.

Reply:
I could go either way. I guess I'd lean towards removal.

Reply:
It'd certainly be nice to have a context on all these functions but that's a major refactoring of the library that should be done in a separate PR.

The wait really depends on the rate-limit here, and I don't think it should hang forever. Adding a timeout seems overly complex and I'd remove that. It might be possible to calculate some upper limit based on the rate-limit used and the max chunk size perhaps.

Reply:
To confirm my understanding: 
1. Remove the timeout from the context.
2. Allow the ops to wait indefinitely, with no timeout.

I'm ultimately only concerned with throttling cache operations against my local store, so I think that's fine for me. However, I haven't looked at the serving code much, is there a chance that this affects HTTP handling?

Reply:
Correct, no custom timeout. This can be implemented some day when contexts are added to all store operations.

This shouldn't affect HTTP stores at all, since we're not really limiting the data-rate, but the speed at which whole chunks can be read/stored. So if the rate-limiter allows one chunk to be accessed, that will be transferred at max speed which means the HTTP timeouts apply like normal. Chunk downloading is limited to the rate-limit on average, not for individual chunks.

Reply:
I added a test with the HttpHandler. There is a way to reproduce an error.

```
goroutine 125 [select]:
golang.org/x/time/rate.(*Limiter).wait(0xc0001df360, {0xe27338, 0x13ceb80}, 0x1, {0x0?, 0x0?, 0x139e0c0?}, 0xd56f30)
        /home/bstrausser/go/pkg/mod/golang.org/x/time@v0.5.0/rate/rate.go:285 +0x3e5
golang.org/x/time/rate.(*Limiter).WaitN(0x79e845e27a40b76?, {0xe27338, 0x13ceb80}, 0x269063b6f56504d5?)
        /home/bstrausser/go/pkg/mod/golang.org/x/time@v0.5.0/rate/rate.go:248 +0x50
github.com/folbricht/desync.RateLimitedStore.HasChunk({{0xe27dc0, 0xc0003f2b00}, 0xc0001df360, {0x3fc999999999999a, 0x1}}, {0x76, 0xb, 0xa4, 0x27, 0x5e, ...})
        /home/bstrausser/Git/desync/ratelimitstore.go:59 +0x7e
github.com/folbricht/desync.HTTPHandler.head({{{0xceccfb, 0x5}, 0x1, {0x0, 0x0}}, {0xe273e0, 0xc00040cd50}, 0x0, {0xc0003f0990, 0x1, ...}, ...}, ...)
        /home/bstrausser/Git/desync/httphandler.go:77 +0x4e
github.com/folbricht/desync.HTTPHandler.ServeHTTP({{{0xceccfb, 0x5}, 0x1, {0x0, 0x0}}, {0xe273e0, 0xc00040cd50}, 0x0, {0xc0003f0990, 0x1, ...}, ...}, ...)
        /home/bstrausser/Git/desync/httphandler.go:47 +0x266
net/http.serverHandler.ServeHTTP({0xc001bda720?}, {0xe25808?, 0xc00061c2a0?}, 0x6?)
        /usr/local/go/src/net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc0006a0c60, {0xe275d8, 0xc00040d1a0})
        /usr/local/go/src/net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 9
        /usr/local/go/src/net/http/server.go:3086 +0x5cb
```

Reply:
That's just part of the stacktrace. Is there a repro on your branch? What test can I run to see it?

Reply:
When I went back to look I couldn't repro.